src/lib/Transform.ts

```diff
@@ -185,37 +195,43 @@ export function batch<T>(
 
 	readable.forEach(
 		async (v: T): Promise<void> => {
-			throwIfThrowable(consumeEarlyEndError());
-
+			startFlushTimeout();
 			queue.push(v);
+			let flushFailureError: Error | undefined;
+			let earlyFlushError: Error | undefined;
 
 			if (queue.length >= maxBatchSize) {
-				// backpressure
-				await flush();
-			} else if (queue.length >= minBatchSize && !pendingWrite) {
+				try {
+					// backpressure
+					earlyFlushError = await settleEarlyFlush();
+					await flush();
+				} catch (e) {
+					flushFailureError = e;
+				}
+			} else if (!pendingWrite) {
 				// no backpressure yet (until new queue fills to maxBatchSize)
-				earlyFlush();
+				swallowErrors(earlyFlush());
 			}
 
-			if (queue.length && flushTimeout !== undefined && !pendingWrite) {
-				if (timer !== undefined) {
-					clearTimeout(timer);
-				}
-				timer = setTimeout(earlyFlush, flushTimeout);
-			}
+			throwIfThrowable(earlyFlushError);
```

Comment:
Whoa, this whole thing got quite a bit more complex than I initially thought...
I must admit I find it a bit hard to follow everything. There are quite a few different branches through the code, and it's hard to even see whether we covered all possible combinations in our tests.
I'm wondering whether it may help to model everything as a statemachine. But that's another rewrite, and I thought we were pretty close yesterday already...

Anyway, I have 10 minutes left to try and debug the tests. Back to work :)